### ISSUE TITLE
Uniplate Derive support for Structs and Unions

### ISSUE THREAD

**Issue by gskorokhod** [conjure-cp/conjure-oxide#261]
Currently, the derive macro only supports enums (see #259).
Before we can properly release uniplate as a separate crate, we should probably implement it for structs and unions

**Comment by niklasdewally** [comment]
I am also not too sure I see how they would work with structs - my understanding was that they solve problems to do with sum types (enums), but structs don't have variants?

**Comment by ozgurakgun** [comment]
What if you had a struct nested in one of the variants? So for completeness they need to be supported, I think.

**Comment by niklasdewally** [comment]
> What if you had a struct nested in one of the variants? So for completeness they need to be supported, I think.

So its for cases where we have an enum T containing a struct that itself contains an enum T?

I think in the paper "transient nesting" isn't supported by the base Uniplate derivation? I think later on he discusses making a `Uniplate<T>` be defined as `Biplate<T,T>`, which would detect transient children, but this came with other limitations iirc, and we will need biplates first?

**Comment by niklasdewally** [comment]
> Don't know how the derivation will work, but you can see how this can be manually implemented, right?

For a Biplate you derive O(n^2) instances, where n is the number of types in the structure. Each `Biplate<T,U>` tells you how to find a type T from type U. You then walk towards the right type recursively with different biplates. 

For example, the path for a type T that contains a V that contains a U that contains a T you would:

Call universeBi on Biplate<T,T>
Call universeBi on the child V using Biplate<T,V>
Call universeBi on the child  U using Biplate<T,U>
This is a base case and would return the T.

**Comment by niklasdewally** [comment]
Although the paper also seems to suggest that Biplate<T,T> just does nothing and returns itself as its children.

**Comment by ozgurakgun** [comment]
But if everything was a T that should boil down to the same thing?

The situation I am thinking of is: if a variant of an enum T contains a struct S that contains a T. In that case should the T be gives as a child of the enum or not. Not sure what Haskell does in this case though I can check!

**Comment by ozgurakgun** [comment]
Something like

```
data T = A Int | B T T | S D

data D = D T T

test1 = B (A 1) (A 2) -- should return 2 children of type T
test2 = S (D (A 1) (A 2) -- 0 or 2 children of type T
```

**Comment by ozgurakgun** [comment]
The results are in!

t.hs
```
{-# LANGUAGE DeriveDataTypeable #-}

import Data.Data
import Data.Generics.Uniplate.Data

data T = A Int | B T T | S D
          deriving (Show, Eq, Data, Typeable)

data D = D T T
          deriving (Show, Eq, Data, Typeable)


test1 = B (A 1) (A 2) -- should return 2 children of type T
test2 = S (D (A 1) (A 2)) -- 0 or 2 children of type T
```

```
$ ghci t.hs
[1 of 1] Compiling Main             ( t.hs, interpreted )
Ok, one module loaded.
ghci> children test1
[A 1,A 2]
ghci> children test2
[A 1,A 2]
ghci>
```

**Comment by ozgurakgun** [comment]
The plot thickens (biplate now)

```
ghci> children test2 :: [T]
[A 1,A 2]
```

Ok, as expected

```
ghci> childrenBi test2 :: [T]
[S (D (A 1) (A 2))]
```

And turns out Biplate<T,T> and Uniplate<T> are **not** identical.

```
ghci> childrenBi (D (A 1) (A 2)) :: [T]
[A 1,A 2]
```

This is also as expected.

**Comment by niklasdewally** [comment]
We (@gskorokhod, @ozgurakgun and I) should probably discuss this more on Friday- it seems that we need some sort of Biplates to implement Uniplates properly.

An important mistake I made is that children is *maximal substructures of the same type*, not direct children - this explains the examples above.

Oz and I have just met to discuss this - the basic idea we came up with is to do something like this (in psuedo-Haskell):

```hs
data T = A Int | B T T | S D
          deriving (Show, Eq, Data, Typeable)

data D = D T T
          deriving (Show, Eq, Data, Typeable)

instance Uniplate of T where
  uniplate A a = [] -- a is not an algebraic data-type, so no need to check it
  uniplate B a b =  (childrenBi a ::[T]) ++ (childrenBi b ::[T])
  uniplate S d =  (childrenBi d ::[T])
-- ...
```

`childrenBi a ::[T]` means find all `T` typed children of a using a Biplate.

This makes the weird case with `Biplate<T,T>` make sense - if a is not a T, it will find all nested T's using the biplate. If a is a T, it will just return a and not descend deeper.

This is just intuition - there is lots we can do to push this more onto build-time and optimise, etc etc.

Section 5 of the paper describes the derivation of Uniplate and Biplate in terms of a common helper class. This uses a set of combinators that mark whether a field is T, may contain T, or will never contain T.  This is analogous to the intuition shown above, and may be helpful to think about.

**Comment by niklasdewally** [comment]
@gskorokhod this also gives us collections and deeper nesting and so on!

This is the combinators implementation of Uniplate I discussed above : 
https://hackage.haskell.org/package/uniplate-1.6.13/docs/Data-Generics-Uniplate-Direct.html

**Comment by ozgurakgun** [comment]
I think thinking in terms of children or childrenBi can be misleading. Think in terms or uniplate and biplate (as we had on the whiteboard)

You _could_ call biplate on primitive types if you wanted to. I trust the rust compiler to optimise those away :)

We want:

```
uniplate :: (a -> Tree<a>, Tree<a> -> a)

biplate :: (a -> Tree<b>, Tree<b> -> a)
```

For a suitable Tree type. `Tree a = Empty | Leaf a | Branch(Vec<a>)` will probably do. You don't need Empty, strictly speaking, but probably nice to have.

BTW I am loving this pseudo Haskell/Rust hybrid. Sorry none of the pseudocode we write in this thread is valid Haskell nor Rust!

**Comment by niklasdewally** [comment]
A picture of the board for future reference: 
<img width="513" alt="Screenshot 2024-03-12 at 17 32 26" src="https://github.com/conjure-cp/conjure-oxide/assets/44070547/3998ec68-e500-42f4-96c5-47ea179b0e43">

**Comment by niklasdewally** [comment]
This seems important to how trees work!

(Data/Generics/Str.rs)
![Screenshot 2024-03-20 at 21 04 45](https://github.com/conjure-cp/conjure-oxide/assets/44070547/c56d2d1f-b6d6-4015-bdbb-73e772ff8cab)

**Comment by niklasdewally** [comment]
In terms of Tree functions, we just need this and maybe a map.

**Comment by niklasdewally** [comment]
Someone at Stack Overflow did some of uniplate in C# too: https://github.com/benjamin-hodgson/Sawmill

**Comment by ozgurakgun** [comment]
I thought we had decided that we wouldn't need Str exactly as it is implemented in Haskell, instead we would have a Tree that can have vectors as children. Is there a problem with that idea?

**Comment by niklasdewally** [comment]
#der

> I thought we had decided that we wouldn't need Str exactly as it is implemented in Haskell, instead we would have a Tree that can have vectors as children. Is there a problem with that idea?

I agree. 

I still think the strStructure function is needed for us , and the idea of it is basically the same with vectors instead of Two.

**Comment by niklasdewally** [comment]
I am currently doing a writeup on all of this (which I was sending drafts of to you on Teams) and will post it here when done.

**Comment by niklasdewally** [comment]
RFC @gskorokhod  @ozgurakgun  (and anyone else!)

I have wrote a memo on the wiki detailing how uniplates and biplates can be implemented using a tree data-structure.

 https://github.com/conjure-cp/conjure-oxide/wiki/2024%E2%80%9003%E2%80%9021-Implementing-Uniplates-and-Biplates-with-Structure-Preserving-Trees
